src/engine/strat_engine/backstore/metadata.rs

```diff
@@ -196,68 +231,74 @@ impl StaticHeader {
+
+        match (
+            StaticHeader::sigblock_from_buf(buf_loc_1),
+            StaticHeader::sigblock_from_buf(buf_loc_2),
+        ) {
+            (Ok(loc_1), Ok(loc_2)) => {
+                match (loc_1, loc_2) {
+                    (Some(loc_1), Some(_)) => {
+                        return Ok(Some(loc_1));
+                    }
+                    (None, None) => {
+                        return Ok(None);
+                    }
+                    (Some(loc_1), None) => {
+                        // Copy 1 has valid Stratis BDA, copy 2 has no magic, re-write copy 2
+                        BDA::write(f, buf_loc_1, MetadataLocation::Second)?;
+                        return Ok(Some(loc_1));
+                    }
+                    (None, Some(loc_2)) => {
+                        // Copy 2 has valid Stratis BDA, copy 1 has no magic, re-write copy 1
+                        BDA::write(f, buf_loc_2, MetadataLocation::First)?;
+                        return Ok(Some(loc_2));
+                    }
+                }
+            }
+            (Ok(loc_1), Err(_)) => {
+                // Re-write copy 2
+                if loc_1.is_some() {
+                    BDA::write(f, buf_loc_1, MetadataLocation::Second)?;
+                }
+                return Ok(loc_1);
+            }
+            (Err(_), Ok(loc_2)) => {
+                // Re-write copy 1
+                if loc_2.is_some() {
+                    BDA::write(f, buf_loc_2, MetadataLocation::First)?;
+                }
+                return Ok(loc_2);
+            }
+            (Err(_), Err(_)) => {
+                let err_str = "Appeared to be a Stratis device, but no valid sigblock found";
+                return Err(StratisError::Engine(ErrorEnum::Invalid, err_str.into()));
             }
         }
-
-        let err_str = "Appeared to be a Stratis device, but no valid sigblock found";
-        Err(StratisError::Engine(ErrorEnum::Invalid, err_str.into()))
     }
 
     /// Determine the ownership of a device.
     /// If the device is owned by Stratis, return its device UUID.
     pub fn determine_ownership<F>(f: &mut F) -> StratisResult<DevOwnership>
     where
-        F: Read + Seek,
+        F: Read + Seek + SyncAll,
     {
-        #![allow(unused_io_amount)]
-
-        f.seek(SeekFrom::Start(0))?;
-        let mut buf = [0u8; _BDA_STATIC_HDR_SIZE];
-
-        // FIXME: See https://github.com/stratis-storage/stratisd/pull/476
-        f.read(&mut buf)?;
-
         // Using setup() as a test of ownership sets a high bar. It is
         // not sufficient to have STRAT_MAGIC to be considered "Ours",
         // it must also have correct CRC, no weird stuff in fields,
         // etc!
-        match StaticHeader::setup_from_buf(&buf) {
+        match StaticHeader::setup(f) {
             Ok(Some(sh)) => Ok(DevOwnership::Ours(sh.pool_uuid, sh.dev_uuid)),
             Ok(None) => {
+                let buf = BDA::read(f)?;
```

Comment:
This is the interesting bit:

```determine_ownership``` now reads the BDA twice, if no Stratis heade ris found, where before it only read once. If computation is structured in the way it currently is, that is necessary, because ```StaticHeader::setup```, since it does its own repair work, must have access to the actual file.

There are obviously a number of ways to restructure the computation, as:
* ```StaticHeader:;setup``` can return instructions on how to fix stuff up and therefore only need to take a buffer argument.
* ```StaticHeader::setup``` can return the buffer that is currently on the disk as a result of its actions, whatever that is, even if it took no action. This too means it can take the allready read buffer, not the file.
* etc.

Or the choice can be made to accept the double read.

Reply:
This code path is quite a bit different in https://github.com/stratis-storage/stratisd/pull/936 I suggest we just leave as is.